### PR TITLE
Add a new code gallery program: Information density based mesh refinement for inverse problems.

### PR DESCRIPTION
**Pull request by bangerth**
Having spend a couple of days cleaning up the code gallery hopefully earns me the right to submit a new program and get it reviewed :-)

### CODE REVIEW

File: information_based_mesh_refinement/mesh_refinement.cc [hunk 1]
```diff
@@ -0,0 +1,890 @@
...
+// "real" detector locations are, but it does not introduce a large
+// error to do this.
+template <int dim>
+void InformationDensityMeshRefinement<dim>::bounce_measurement_points_to_cell_centers ()
+{
+  detector_locations_on_mesh = detector_locations;
+  for (auto &p : detector_locations_on_mesh)
+    {
+      for (const auto &cell : triangulation.active_cell_iterators())
+	if (cell->point_inside (p))
```
Comment (Rombur) [inline information_based_mesh_refinement/mesh_refinement.cc:337]: Indentation is weird

File: information_based_mesh_refinement/mesh_refinement.cc [hunk 2]
```diff
@@ -0,0 +1,890 @@
...
+
+        for (unsigned int q_point=0; q_point<n_q_points; ++q_point)
+          for (unsigned int i=0; i<dofs_per_cell; ++i)
+            for (unsigned int j=0; j<dofs_per_cell; ++j)
+              cell_matrix(i,j) += (fe_values.shape_grad (i,q_point) *
+                                   fe_values.shape_grad(j,q_point)
+				   +
+				   fe_values.shape_value(i,q_point) *
+				   (velocity * fe_values.shape_grad(j,q_point))) *
+                                  fe_values.JxW(q_point);
```
Comment (Rombur) [inline information_based_mesh_refinement/mesh_refinement.cc:573]: Weird indentation

File: information_based_mesh_refinement/mesh_refinement.cc [hunk 3]
```diff
@@ -0,0 +1,890 @@
...
+                          // contribution to the information content on all
+                          // cells of the mesh by taking into account the
+                          // detector locations. We add these into global
+                          // objects, so we have to guard access to the
+                          // global object:
+                          static std::mutex m;
+                          std::lock_guard<std::mutex> g(m);
+                   
+
+                          // do we need to scale with the measure of the cell here or its square?
```
Comment (Rombur) [inline information_based_mesh_refinement/mesh_refinement.cc:649]: Do you want to keep this comment?

**Review by Rombur — APPROVED** [review]
It looks like you are using two different styles of indentation in the code. Since it's a gallery code, it's fine with me. I just pointed a couple of places where the indentation makes the code harder to read. Also I am pretty sure there is a typo in the name of one the authors.

File: information_based_mesh_refinement/Readme.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+Readme file for "Information density-based mesh refinement"
+===========================================================
+
+@note This program implements the ideas and algorithms described in
+  the paper "Estimating and using information in inverse problems" by
+  Wolfgang Bangerth, Chris R. Johnson, Dennis K. Njeru, and Bart can
```
[the remark this answers is not in the document]
Comment (bangerth) [reply information_based_mesh_refinement/Readme.md]: Good eye, thanks!

**Pull request merged**